### PR TITLE
Update Manga in Expected Period 

### PR DESCRIPTION
**Pull request by quangkieu**
This is mostly to tune down number of fetching updates from source per mange for people that have big collection
This is also lower the pulling load on the source server as they usually paid per request
If we kind of guessed when there would be a new release, it is just wasted if we try to fetch new update before the release cycle

Add Update Interval, Update Cycle column
Add new option for only update manga has next update in expected range
Add new option for setting expected range
Edit update queue logic to use leading expected day range
Edit sync chapter to set next update day after passed following day range
Edit delta to only use unique days from 50 latest chapters, or 20 latest dates (lesser of the two limits)
  minimum to 1 day or default 7 if not enough dates for calculation
Double update interval if missed 10 updates. Double again or minimum 14 days when missed 20 updates
Add cycle to calculate next update time

### CODE REVIEW

**Comment by quangkieu** [comment]
New option for setting expected range

imageimage

Expected interval and update date display in Manga Chapter List
image

I think we could remove the update cycle if not needed

**Comment by quangkieu** [comment]
But in analog engineering, we have to capture in faster half interval of the expected frequency to get the correct frequency. So if manga in 6-days interval then we have to capture every 3-days interval. Should we do that here? It would respond faster to when scan group changes the frequency or new group pick the old manga
I wonder should I have another third double interval?

**Comment by quangkieu** [comment]
I wonder if we could schedule this for next release as seem like IsekaiScan have problem on bandwidth as server keep getting down.
This pull request would somewhat keep the client from abuse on keep track of too much manga and frequently pulling on the server on not scheduled mangas.
If there is concern about missing update, we could just do half interval like suggested above

**Comment by Soitora** [comment]
Could help alleviate sources if this is implemented right, but it looks like you know what you're doing 👍🏻

**Comment by vengiare** [comment]
> Expected interval and update date display in Manga Chapter List

Probably steal the styling from here so it looks better than it currently is.

IMG_20210918_193859

Could also merge two lines into one or something. Dunno which exactly. Maybe period and interval, or interval and cycle.

**Comment by KokaKiwi** [comment]
Would be cool to add a way to set the expected range/period manually for some titles, either when knowing the actual values or to "fix" when the guess is totally wrong (for example, some sources don't date each chapters then the app "guess" on way too large period, which result in almost never updating some titles)

**Comment by quangkieu** [comment]
Version 3
Check Manga in Expected Period Options

* Add Update Interval, Update Cycle column
* Add new option for update only manga has next udpate in expected range
  Add new option for setting expected range
* Edit update queue logic to use leading expected day range
* Edit logic to calculate delta in sync chapter
  Edit sync chapter to set next update day outside leading + following day range
  Edit calDelta to only use unique days from 50 latest timestamped chapters, or 20 latest days
    calDelta only in 1 to 28 days
  Continuous double update interval if missed 10 updates of previous delta. Stop if effecive delta >=56
  Add cycle to calculate next update time
* Display Expected Period, Estimated/Modfied/Effective Intervals, Curent Cycle
* Allow modify interval, and multiplier to calculate Effective Interval

Reorder and increase update interval when late

Add more comments
Add empty chapter check
Move skip-check for in expected time range earlier
Double continuous if missed more than 9 cycles on previous delta


* Add Interval tag to manga based on intervals
  Late Tag if missed first 9-cycles
  Dropped/Axed Tag if missed second 9-cycles
* Add Modied Tag if manga has custom setting

**Comment by quangkieu** [comment]
Set custom update interval
Generate Interval Tag: Modified, Late?, Dropped/Axed?
These tags could localize per language

imageimage
image

**Comment by Soitora** [comment]
It's (probably) great, but it needs to be user friendly too
Also not sure about occupying the tags for this

**Comment by quangkieu** [comment]
I wonder if you have any suggestions for the UX/UI as I think it very simple
As for the tags, my main goal is allow for searching for those "Dropped?" Tag to either migrate to different sources or remove

**Comment by Soitora** [comment]
I don't think anything should be in the tags that is not a genre, border-line not even demographics. (Although being able to search for it wouldn't be that bad, but I don't think this is the method to do so)

As for the other part, I'm not completely sure how to do it, the popup for editing could maybe be like the Tracker edit and be a fullscreen, that allows for better descriptions and such.

Here's how it *could* look on the manga page:
image
More info than that I think is too much, better to move the more detailed stuff somewhere else.
The editing of it could be done through the overflow menu at the top like the other changes

This is just a suggestion from me but I'm fairly certain this is the most amount of extra clutter people would be willing to accept

**Comment by quangkieu** [comment]
Then I wonder how to display that we would check every 28 days (4 weeks, 1 month) if average value greater than 28?
Also, how about the modifier number as we double the interval if no release in 9 cycles of previous value until we over 56 (2 month)
And we need to display cycle too as user might misunderstand why no check today
I think I could move the search for late/drop as view filter as that is using modifier number

**Comment by Jays2Kings** [comment]
> imageimage

I'd change this to a more friendly wording, like "Grace period for expected updates" and change "leading/following" to "Check `x` days before/after". Just having numbers with no time unit is hard to understand what the numbers mean

> Expected interval and update date display in Manga Chapter List image
> 
> I think we could remove the update cycle if not needed

rather than showing this so front and center, this should 
1) show between tags and the number of chapters 
2) only show if the user expands the description via the chevron
3) this should never show if the user disabled "only update in expected period"

Also I'm not sure at all what update cycle is so I cant say if its relevant to users or not

> Then I wonder how to display that we would check every 28 days (4 weeks, 1 month) if average value greater than 28?
Also, how about the modifier number as we double the interval if no release in 10 cycles of previous value until we over 28 (1 month)

Sometimes you dont need to know how the spaghetti is made. Users dont need to know it checks every 28 days, just knowing when its going to updated. Showing it like "next expected update in" like soi has it should be fine. If they're currently in that expected period the text can cange to "now-2 days" because you dont need to show how long in the expected period they are in. They'll find out next time they're outside the expected period.

**Comment by quangkieu** [comment]
> _2¢ from a random contributor_
> 
> Is there a reason why the UI for this feature couldn't just be an `Automatic` entry under `Update Frequency`? The whole point of this feature is intelligent updates per library entry, so like Jay said why show users how the spaghetti is made.
> 
> Not having any additional UI also makes it so that the implementation can be tuned/changed more easily over time.

@h-hyuuga, the "Update frequency" is using to set the time when update service to run. This feature is just to change the behavior of that service to only check a subset of the manga in library to update

**Comment by quangkieu** [comment]
New image

<details>
<summary>Details</summary>
image
image
image
image
</details>

I could not add new string resource without all the string map to wrong values.

**Comment by quangkieu** [comment]
I am working on another feature but I would have to add a few more resource strings. It would work on 3 more but if I add the 4th one, all the strings are mapping to wrong value. I wonder if we are reaching the string limit?

**Comment by Soitora** [comment]
1. Having the text be "Estimated update: $"
2. Having the update text be in a darker color, like the date subtitles
image

3. This part can be reworded but I don't know what to write instead
image

4. Surely there has to be a better way to write this for userfriendliness
image

4. This can probably be reworded to "Only use smart updates"
image

6. This can probably be reworded to "Smart updates" or something
image

Good job so far
I got more to comment on but not right now

**Comment by quangkieu** [comment]
Default estimate interval
image

When user set custom interval
image

Setting
image

New filter

**Comment by quangkieu** [comment]
As most of the feedback are more in the text description, I think we could do that in another pull request so insider could test in the weekly release about the logic first?

**Comment by FourTOne5** [comment]
Why not seperate the filters by category? the current one will be category less and there will estimated update status and estimated update interval or something similar.

**Comment by quangkieu** [comment]
Split filters into group
image

**Comment by Cologler** [comment]
Great work! I am wondering if this also works on some mangas without an update date?

Screenshot_20211225-005520_Tachiyomi

**Comment by quangkieu** [comment]
> Great work! I am wondering if this also works on some mangas without an update date?
> 
> Screenshot_20211225-005520_Tachiyomi

So right now, we calculate using upload dates, without that, it would use check dates for calculation, even the hybrid of them when server remove upload date on old release. But if you just start check that manga, then all the dates would be the same, so the default 7 days kick in

**Comment by quangkieu** [comment]
rebased and ready for merge

**Comment by arkon** [comment]
I'll review this soon-ish, but I expect that the UX will need to be heavily simplified before making it out to a stable release.

**Comment by CrepeTF** [comment]
This is my suggestion for the UX:
https://www.figma.com/file/oPT2GHeIbZrIVSjvOJ7Cyz/Manga-details?node-id=0%3A1

**Comment by FourTOne5** [comment]
> These are my suggestions for the UX: https://www.figma.com/file/oPT2GHeIbZrIVSjvOJ7Cyz/Manga-details?node-id=0%3A1

the 3rd and 2nd one looks good.

**Comment by Soitora** [comment]
Since you didn't post picture after you tweaked UI
I still believe we'd need to change a bit of UX after merge though, make it even more easy to understand
image

**Comment by quangkieu** [comment]
> Since you didn't post picture after you tweaked UI
> I still believe we'd need to change a bit of UX after merge though, make it even more easy to understand

Do you have other manga that show thí 1-1?
I only saw this

**Comment by quangkieu** [comment]
Yeah we could tweak it as i am not familiar with UX UI in Android development much

**Comment by hyoretsu** [comment]
Not sure if this has been perfected yet, but "Expected: 3~4 days"

**Comment by suhan-paradkar** [comment]
> I think best course of action is merging if code works and then "beautify" and tweak it after

Yes..  since it can greatly reduce load on the sources...

**Comment by quangkieu** [comment]
So I submit this for a long time ago. It works. But none approve for the merge and now 4 months after the code base moved a lots already. The changes in architecture, the async call. I tried to rebase and it took me 2 months but it got a small bug. I am not sure how to keep up with all these big architectural changes

**Comment by arkon** [comment]
Sorry, I was waiting for the UX to be improved first but it probably makes more sense to just merge first and try to iterate on it as needed.

I think most of the major database/UI migrations have been completed by now, but it'd probably make sense to try to redo the database/chapter source sync bits first, then the preferences/UI separately.

**Comment by jferdous70** [comment]
I hope this doesn't get closed like the other PRs due to compose UI changes

**Comment by Soitora** [comment]
> I hope this doesn't get closed like the other PRs due to compose UI changes 

I don't think you need to worry about that part

**Comment by AntsyLich** [comment]
You should make the PR a draft until it's ready

**Comment by Soitora** [comment]
> You should make the PR a draft until it's ready

Hasn't it been considered "ready" for a while?
Just commits to main branch making this needing to be modified to be up-to-date

**Comment by arkon** [comment]
I think it'd make more sense to start with a smaller PR that only includes the data-layer changes (migrations, writing of the data). That'd give us some leeway in figuring out the proper UI/UX for it, especially given that the UI rewrites for the parts that you're touching are still in progress.

Things to consider:
- The interval data could be a separate table (with the manga ID as the key).
- The interactor bits can be a separate class. It'd be worth having unit tests too for it too, since the logic is somewhat complex.

**Comment by quangkieu** [comment]
Change the design to just show interval in manga info section

**Comment by AntsyLich** [comment]
The normal icon is used by trackers better to change it

**Comment by CrepeTF** [comment]
I suggest making use of a variation of the Forward 5/30 icons

**Comment by quangkieu** [comment]
image
image
Split into multiple commit for easy preview
Change icon
Update most dialogs to use new composable classes

**Comment by arkon** [comment]
@quangkieu Please pull out the data-layer changes (migrations, writing of the data) into a separate smaller PR first.

Things to consider:

- The interval data could be a separate table (with the manga ID as the key).
- The interactor bits can be a separate class. It'd be worth having unit tests too for it too, since the logic is somewhat complex.

**Comment by mickae1** [comment]
Hi, I'm waiting for this awesome feature, 
I have a large collection and it takes time to update.

@quangkieu good luck !

**Comment by mickae1** [comment]
Some Website offer give the date of the next update, maybe tachiyomi could use this information to improve the update.

**Comment by mickae1** [comment]
@arkon it doesn't looks like that @quangkieu has the time to work on it.

This feature is very important, more than other in my opinion. Extension has to be updated frequently.. With this feature, this will help a lot.

Could you accept this pull request ?
And maybe later make some change ?

**Comment by mickae1** [comment]
@arkon please :)

**Comment by AntsyLich** [comment]
@mickae1 It's impossible to merge this PR as it has conflicts. The OP needs to resolve the conflicts. He might as well do what arkon asked so it's easier for him to manage and arkon to review

**Comment by quangkieu** [comment]
Split into 3 pull request
https://github.com/tachiyomiorg/tachiyomi/pull/9398
https://github.com/tachiyomiorg/tachiyomi/pull/9399
https://github.com/tachiyomiorg/tachiyomi/pull/5734 (and this)

**Comment by quangkieu** [comment]
I got this problem with Android Studio. All build are success but Gradle sync would fail and lots of reference for helper could not find

<details>
<summary>Details</summary>

image
Caused by: java.lang.IllegalStateException: Artifacts of dependency org.jetbrains.kotlinx:kotlinx-coroutines-core-metadata:1.4.1 is built by old Kotlin Gradle Plugin and can't be consumed in this way

</details>

**Comment by arkon** [comment]
I ran into that at one point too and invalidating/clearing the Android Studio cache helped.

**Comment by quangkieu** [comment]
I still got same error after IDE clear cache or manually remove any gradle folder on machine. I wonder if we could remove org.jetbrains.kotlinx:kotlinx-coroutines-core-metadata:1.4.1 in source-* project somehow? I am not familiar with android build tool to remove that

**Comment by arkon** [comment]
I'm actually not even really sure where it's coming from, we use 1.6.4.

I suspect it may be related to AGP, so you could try reverting [this change](https://github.com/tachiyomiorg/tachiyomi/commit/97e6f1ea9a4b31d9b2cee0ffc4f032f3ba6d6bc7) locally to see if that helps.

**Comment by quangkieu** [comment]
I found problem. It is from bump to kotlin_version = "1.8.20" from kotlin_version = "1.8.10" commit

**Comment by arkon** [comment]
Someone else on Discord also narrowed it down to the same thing, so f317193becb5646b3b7eca794311745cc2ba29e0 should help.

**Comment by Soitora** [comment]
Seeing as you still push to this, maybe set it to Draft until you feel like it's ready for finalized merge?

**Comment by quangkieu** [comment]
I push new rebased on master as master has conflict

**Comment by khan-de-code** [comment]
Would this be the thread to propose a UI/UX addition? So far this thread seems to focus on altering the user of stale download sources from within the media view, but the only way I've seen to view all of your stale media en-masse is via a somewhat obscure filter, which would mean the user still has to remember to check the filter occasionally for stale media.

I proposed in #9489 before finding this thread that each category could have a subsection where all stale media are automatically moved to, so that they are still visible in the category, but clearly separated from the continuously updated ones.

My vision for the visual representation of the stale subsection is that in each category, should there be at least one stale media identified, a dividing line appears below all of your current media with the media tagged as stale residing below that section. Perhaps the dividing line could also have the word Stale (left aligned, middle aligned, whatever looks best) in it, like so:

╭――――――――――――――――――╮
│                                                                 
│ Media                                                      
│                                                                 
╰――――――――――――――――――╯
Stale -------------------------------------------------------------
╭――――――――――――――――――╮
│                                                                 
│ Stale Media                                             
│                                                                 
╰――――――――――――――――――╯

This would benefit users to quickly identify their stale media at a glance and look for alternate sources, rather than having to remember to check a filter or take another step to enter into a media's page to see that it has become stale.

Another idea to augment my above suggestion would be to have a small bar underneath the category tabs that would help draw attention to the stale media section via being an eye grab and possibly being clickable such that when clicked, scrolls your screen down to the stale media section.

**Comment by quangkieu** [comment]
@khan-de-code I do like your idea but I would say finish the current pull request, finish the logic to identify stale and all background. Then we could open another for the UI/UX that separate them out. I think I would then switch back to my original proposal that add Late/Drop tag to the comic. Then it would work for both filter and frontend design. I did got push back on that idea back then

**Comment by arkon** [comment]
Sorry for the conflicts, I did some minor cleanup on top of the work from #9399. You should be good to rebase this on top of that.

**Comment by arkon** [comment]
Also something I missed in #9399 -- why are the lead/follow days defaulted to 1 day, despite the "Default" in the dialog corresponding with 0 days?

**Comment by quangkieu** [comment]
I think I might mess-up something in those rebase, I remember I keep it from 0-28 instead of Default,1-28.

**Comment by quangkieu** [comment]
I move the filter to another pull request so this could be close easier and being used by everyone earlier. As the filter got so many designs, but we would need final agreeable design.

**Comment by mickae1** [comment]
> I move the filter to another pull request so this could be close easier and being used by everyone earlier. As the filter got so many designs, but we would need final agreeable design.

Can't wait to test your work :)

**Comment by quangkieu** [comment]
Is there any delay on this?

**Comment by arkon** [comment]
I'll follow up on this and #9619 within the next few days.

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt [hunk 1]
```diff
@@ -423,6 +446,21 @@ class MangaInfoScreenModel(
 
     // Manga info - end
 
+    fun setEstimateInterval(manga: Manga, interval: Int) {
+        val calculateInterval = when (interval) {
+            0 -> if (manga.calculateInterval < 0) 0 else manga.calculateInterval
```
[the remark this answers is not in the document]
Comment (quangkieu) [reply app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt:451]: In here, I reset interval 0 default to trigger recalculation. We only reset if interval is custom, which is negative

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreen.kt [hunk 2]
```diff
@@ -207,6 +210,15 @@ class MangaScreen(
                     LoadingScreen(Modifier.systemBarsPadding())
                 }
             }
+            is MangaInfoScreenModel.Dialog.SetMangaInterval -> {
+                SetIntervalDialog(
+                    intervalInterval = if (dialog.manga.calculateInterval < 0) -dialog.manga.calculateInterval else 0,
```
[the remark this answers is not in the document]
Comment (quangkieu) [reply app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreen.kt:215]: I am using negative interval for custom interval that user set. If interval calculated, we keep it 0/ default. Else set to absolute value of the above custom interval so user could change easyly

File: app/src/main/java/eu/kanade/tachiyomi/data/backup/BackupRestorer.kt [hunk 3]
```diff
@@ -128,6 +143,9 @@ class BackupRestorer(
                 // Fetch rest of manga information
                 restoreNewManga(updatedManga, chapters, categories, history, tracks, backupCategories)
             }
+            val upManga = mangaRepository.getMangaByUrlAndSourceId(manga.url, manga.source)!!
```
[the remark this answers is not in the document]
Comment (quangkieu) [reply app/src/main/java/eu/kanade/tachiyomi/data/backup/BackupRestorer.kt:146]: Ah, so right now, we do not store interval or next update data in the backup as it is not needed.

**Comment by quangkieu** [comment]
Ready for review

File: app/src/main/java/eu/kanade/presentation/manga/components/MangaInfoHeader.kt [hunk 4]
```diff
@@ -183,6 +188,20 @@ fun MangaActionRow(
             onClick = onAddToLibraryClicked,
             onLongClick = onEditCategory,
         )
+        if (onEditIntervalClicked != null && interval != null) {
+            MangaActionButton(
+                title =
+                if (interval.first == interval.second) {
+                    pluralStringResource(id = R.plurals.interval_day, count = interval.second)
```
Comment (quangkieu) [inline app/src/main/java/eu/kanade/presentation/manga/components/MangaInfoHeader.kt:195]: @arkon  There is a missing " ,interval.second)" here. I fixed on my branch

**Comment by quangkieu** [comment]
Replace as there was error on merge

- https://github.com/tachiyomiorg/tachiyomi/pull/9751

**Pull request merged**